// rxjs-marbles/expect.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var args_1 = require("./args");
var assert_1 = require("./assert");
var Expect = (function () {
    function Expect(actual, helpers, unsubscription) {
        this.actual = actual;
        this.helpers = helpers;
        this.unsubscription = unsubscription;
    }
    Expect.prototype.toBeObservable = function (expected, values, error) {
        var _a = this, actual = _a.actual, helpers = _a.helpers, unsubscription = _a.unsubscription;
        if (typeof expected === "string") {
            helpers.expectObservable(actual, unsubscription).toBe(expected, values, error);
        }
        else {
            assert_1.assertArgs(expected);
            var _b = expected[args_1.argsSymbol], error_1 = _b.error, marbles = _b.marbles, values_1 = _b.values;
            helpers.expectObservable(actual, unsubscription).toBe(marbles, values_1, error_1);
        }
    };
    Expect.prototype.toHaveSubscriptions = function (expected) {
        var _a = this, actual = _a.actual, helpers = _a.helpers;
        assert_1.assertSubscriptions(actual);
        var subscriptions = actual.subscriptions;
        helpers.expectSubscriptions(subscriptions).toBe(expected);
    };
    return Expect;
}());
exports.Expect = Expect;
